src/mcts/search.cc

```diff
@@ -609,191 +338,505 @@ void Search::UpdateRemainingMoves() {
+    if (AddNodeToCompute(node, false)) {
+      // Make it return 0 to make it not use the slot, so that the function
+      // tries hard to find something to cache even among unpopular moves.
+      // In practice that slows things down a lot though, as it's not always
+      // easy to find what to cache.
+      return 1;
+    }
+    return 1;
+  }
+
+  // If it's a node in progress of expansion or is terminal, not prefetching.
+  if (!node->HasChildren()) return 0;
+
+  // Populate all subnodes and their scores.
+  typedef std::pair<float, Node*> ScoredNode;
+  std::vector<ScoredNode> scores;
+  float factor =
+      search_->kCpuct * std::sqrt(std::max(node->GetChildrenVisits(), 1u));
+  // FPU reduction is not taken into account.
+  const float parent_q = -node->GetQ(0, search_->kExtraVirtualLoss);
+  for (Node* iter : node->Children()) {
+    if (iter->GetP() == 0.0f) continue;
+    // Flipping sign of a score to be able to easily sort.
+    scores.emplace_back(-factor * iter->GetU() -
+                            iter->GetQ(parent_q, search_->kExtraVirtualLoss),
+                        iter);
+  }
+
+  size_t first_unsorted_index = 0;
+  int total_budget_spent = 0;
+  int budget_to_spend = budget;  // Initializing for the case there's only
+                                 // on child.
+  for (size_t i = 0; i < scores.size(); ++i) {
+    if (budget <= 0) break;
+
+    // Sort next chunk of a vector. 3 of a time. Most of the times it's fine.
+    if (first_unsorted_index != scores.size() &&
+        i + 2 >= first_unsorted_index) {
+      const int new_unsorted_index =
+          std::min(scores.size(), budget < 2 ? first_unsorted_index + 2
+                                             : first_unsorted_index + 3);
+      std::partial_sort(scores.begin() + first_unsorted_index,
+                        scores.begin() + new_unsorted_index, scores.end());
+      first_unsorted_index = new_unsorted_index;
+    }
+
+    Node* n = scores[i].second;
+    // Last node gets the same budget as prev-to-last node.
+    if (i != scores.size() - 1) {
+      // Sign of the score was flipped for sorting, flipping back.
+      const float next_score = -scores[i + 1].first;
+      const float q = n->GetQ(-parent_q, search_->kExtraVirtualLoss);
+      if (next_score > q) {
+        budget_to_spend = std::min(
+            budget,
+            int(n->GetP() * factor / (next_score - q) - n->GetNStarted()) + 1);
+      } else {
+        budget_to_spend = budget;
+      }
+    }
+    history_.Append(n->GetMove());
+    const int budget_spent = PrefetchIntoCache(n, budget_to_spend);
+    history_.Pop();
+    budget -= budget_spent;
+    total_budget_spent += budget_spent;
+  }
+  return total_budget_spent;
 }
 
-void Search::Wait() {
-  Mutex::Lock lock(threads_mutex_);
-  while (!threads_.empty()) {
-    threads_.back().join();
-    threads_.pop_back();
+// 4. Run NN computation.
+// ~~~~~~~~~~~~~~~~~~~~~~
+void SearchWorker::RunNNComputation() {
```

Comment:
Although I like the simplicity of SearchWorker::ExecuteOneIteration, surely this one liner can be merely inlined directly into SearchWorker::ExecuteOneIteration?